src/cdk/a11y/input-modality/input-modality-detector.ts

```diff
@@ -0,0 +1,187 @@
+  /** Keys to ignore when detecting keyboard input modality. */
+  ignoreKeys?: string[];
+}
+
+/**
+ * Injectable options for the InputModalityDetector. These are shallowly merged with the default
+ * options.
+ */
+export const INPUT_MODALITY_DETECTOR_OPTIONS =
+  new InjectionToken<InputModalityDetectorOptions>('cdk-input-modality-detector-options');
+
+/**
+ * Default options for the InputModalityDetector.
+ *
+ * Modifier keys are ignored by default (i.e. when pressed won't cause the service to detect
+ * keyboard input modality) for two reasons:
+ *
+ * 1. Modifier keys are commonly used with mouse to perform actions such as 'right click' or 'open
+ *    in new tab', and are thus less representative of actual keyboard interaction.
+ * 2. VoiceOver triggers some keyboard events when linearly navigating with Control + Option (but
+ *    confusingly not with Caps Lock). Thus, to have parity with other screen readers, we ignore
+ *    these keys so as to not update the input modality.
+ */
+export const INPUT_MODALITY_DETECTOR_DEFAULT_OPTIONS: InputModalityDetectorOptions = {
+  ignoreKeys: ['Alt', 'Control', 'Meta', 'Shift'],
+};
+
+/**
+ * The amount of time needed to pass after a touchstart event in order for a subsequent mousedown
+ * event to be attributed as mouse and not touch.
+ */
+export const TIME_AFTER_TOUCH_MS = 650;
+
+/**
+ * Event listener options that enable capturing and also mark the listener as passive if the browser
+ * supports it.
+ */
+const opts = normalizePassiveListenerOptions({
+  passive: true,
+  capture: true,
+});
+
+/**
+ * Service that detects the user's input modality.
+ *
+ * Note: This service will not update the input modality when a user is navigating with a screen
+ * reader (e.g. linear navigation with VoiceOver, object navigation / browse mode with NVDA, virtual
+ * PC cursor mode with JAWS). This is in part due to technical limitations (i.e. keyboard events
+ * do not fire as expected in these modes) but is also arguably the correct behavior. Navigating
+ * with a screen reader is akin to visually scanning a page, and should not be interpreted as actual
+ * user input interaction.
+ *
+ * When a user is not navigating but *interacting* with a screen reader, this service's behavior is
+ * largely undefined and depends on the events fired. For example, in VoiceOver, no keyboard events
+ * are fired when performing an element's default action via Caps Lock + Space, thus no input
+ * modality is detected.
+ */
+@Injectable({ providedIn: 'root' })
+export class InputModalityDetector implements OnDestroy {
+  /** Emits when the input modality changes. */
+  readonly inputModalityChange: Observable<InputModality>;
+
+  /** Returns the last detected / current input modality. */
+  get inputModality(): InputModality {
+    return this._inputModality;
+  }
+
+  /** The underlying ReplaySubject that emits input modality changes. */
+  private readonly _change = new ReplaySubject<InputModality>(1);
```

Comment:
You could replace the `ReplaySubject` with a `BehaviorSubject` which has identical behavior, but it also allows you to read the latest value which can then be used for the `inputModality` getter.

Reply:
This was actually my previous implementation, but I ended switching to a `ReplaySubject` because I didn't like how the `BehaviorSubject` emitted `null` at the start. My thinking was that `inputModalityChange` should only emit when the input modality changes, and I didn't want subscribers to have to filter out the first `null` value in their logic. WDYT?

Reply:
You could add `.pipe(skip(1))` when exposing the subject publicly. It's not a bit deal to leave it as it is though.

Reply:
Okay, I don't feel strongly so I'll switch to the `BehaviorSubject`.